### DISCUSSION TITLE
Parameter or Property Placeholder in Base Address

### DISCUSSION THREAD

**Discussion by joshua5822** [canton7/RestEase#194] [Q&A] [answered]
I've been trying to figure out how to get a base address like "http://api{cluster}.site.com" to work with using the HttpClientFactory extension package. The problem I've ran into is it validates the string using Uri and thus fails as it won't allow curly braces inside the hostname. I was able to get around it by using the base address attribute on the interface, which doesn't use URI parsing, but then I have the issue of how to use a different URL for different environments. 
Ex: "http://api{cluster}.prod.site.com"  "http://api{cluster}.stage.site.com" 

Any suggestions? I do suppose I could use another placeholder property, but that only works in the cases where the URL always maintains the same pattern between all environments, and wouldn't really work if one of the environment's was using localhost.

**Comment by canton7** [comment, marked as answer]
I think you've found all the relevant features. If you pass a base address to RestClient / the HttpClientFactoey equivalent, that gets set on the HttpClient, which needs a Uri, so there's no way around that.

The only other thought I have is to sort this all out in a RequestInterceptor / HttpMessageHandler. You'll need to add some dummy domain to get it past HttpClient, but then you can strip it off and replace it with whatever. You've got access to the IRequestInfo so you can grab any path properties etc, and do stuff with them yourself. Another option is to subclass Requester and override the stuff which constructs the full path, but that's not guaranteed to be stable between releases.

If you can come up with any more general features which would support your case in a near way, I'm open to suggestions.

**Reply by joshua5822** [reply]
For now I've gone with the BaseAddress attribute and using two Path Properties to set the parts of the url. Its working well in production so far, without any reported issues by clients or in our logs. The code before was still using a singleton httpclient, so overall this is an improvement.

**Reply by canton7** [reply]
That's good to hear!
